In [1]:
import pandas as pd
import numpy as np

import anndata as ad

In [2]:
# reading the .h5ad file from the data directory
file_path = 'data/K562_gwps_normalized_bulk_01.h5ad'

adata = ad.read_h5ad(file_path)

print(adata)

AnnData object with n_obs × n_vars = 11258 × 8248
    obs: 'UMI_count_unfiltered', 'num_cells_unfiltered', 'num_cells_filtered', 'control_expr', 'fold_expr', 'pct_expr', 'core_control', 'mean_leverage_score', 'std_leverage_score', 'energy_test_p_value', 'anderson_darling_counts', 'mann_whitney_counts', 'z_gemgroup_UMI', 'mitopercent', 'TE_ratio', 'cnv_score_z'
    var: 'gene_name', 'mean', 'std', 'cv', 'in_matrix', 'gini', 'clean_mean', 'clean_std', 'clean_cv'


In [3]:
adata.to_df()

gene_id                            ENSG00000237491  ENSG00000228794  \
gene_transcript                                                       
0_A1BG_P1_ENSG00000121410                 0.028642         0.046873   
10000_ZBTB46_P1P2_ENSG00000130584         0.055915        -0.058623   
10001_ZBTB47_P1P2_ENSG00000114853         0.030780         0.038029   
10002_ZBTB48_P1P2_ENSG00000204859        -0.104988        -0.168258   
10003_ZBTB49_P1P2_ENSG00000168826         0.140855        -0.035711   
...                                            ...              ...   
9998_ZBTB44_P1P2_ENSG00000196323          0.017919         0.013294   
9999_ZBTB45_P1P2_ENSG00000119574          0.004931         0.061982   
999_C16orf91_P1P2_ENSG00000174109         0.008794         0.116086   
99_ACSL1_P1P2_ENSG00000151726            -0.017405        -0.050810   
9_AAR2_P1P2_ENSG00000131043              -0.015396         0.017742   

gene_id                            ENSG00000188976  ENSG00000187961  \
gene_transcript                                                       
0_A1BG_P1_ENSG00000121410                -0.007014        -0.043553   
10000_ZBTB46_P1P2_ENSG00000130584         0.106205         0.025113   
10001_ZBTB47_P1P2_ENSG00000114853         0.067629        -0.136385   
10002_ZBTB48_P1P2_ENSG00000204859        -0.044452        -0.057588   
10003_ZBTB49_P1P2_ENSG00000168826        -0.089407         0.096020   
...                                            ...              ...   
9998_ZBTB44_P1P2_ENSG00000196323          0.166343         0.202679   
9999_ZBTB45_P1P2_ENSG00000119574          0.024869         0.019066   
999_C16orf91_P1P2_ENSG00000174109        -0.067448        -0.006761   
99_ACSL1_P1P2_ENSG00000151726             0.012079        -0.077181   
9_AAR2_P1P2_ENSG00000131043               0.313340         0.215871   

gene_id                            ENSG00000188290  ENSG00000187608  \
gene_transcript                                                       
0_A1BG_P1_ENSG00000121410                 0.049834         0.055918   
10000_ZBTB46_P1P2_ENSG00000130584         0.272732        -0.114054   
10001_ZBTB47_P1P2_ENSG00000114853        -0.035681        -0.055437   
10002_ZBTB48_P1P2_ENSG00000204859        -0.024554        -0.002767   
10003_ZBTB49_P1P2_ENSG00000168826         0.084299         0.033209   
...                                            ...              ...   
9998_ZBTB44_P1P2_ENSG00000196323          0.026368        -0.026491   
9999_ZBTB45_P1P2_ENSG00000119574         -0.035541        -0.036628   
999_C16orf91_P1P2_ENSG00000174109         0.070513        -0.057997   
99_ACSL1_P1P2_ENSG00000151726            -0.011902        -0.069777   
9_AAR2_P1P2_ENSG00000131043              -0.090744         0.256190   

gene_id                            ENSG00000078808  ENSG00000176022  \
gene_transcript                                                       
0_A1BG_P1_ENSG00000121410                 0.056130        -0.009680   
10000_ZBTB46_P1P2_ENSG00000130584         0.174116        -0.106970   
10001_ZBTB47_P1P2_ENSG00000114853         0.027704        -0.056662   
10002_ZBTB48_P1P2_ENSG00000204859        -0.034466         0.020218   
10003_ZBTB49_P1P2_ENSG00000168826        -0.039355        -0.126822   
...                                            ...              ...   
9998_ZBTB44_P1P2_ENSG00000196323          0.081814         0.051754   
9999_ZBTB45_P1P2_ENSG00000119574          0.021431        -0.088374   
999_C16orf91_P1P2_ENSG00000174109         0.027870        -0.021607   
99_ACSL1_P1P2_ENSG00000151726             0.077219        -0.024646   
9_AAR2_P1P2_ENSG00000131043               0.116805         0.099588   

gene_id                            ENSG00000160087  ENSG00000131584  ...  \
gene_transcript                                                      ...   
0_A1BG_P1_ENSG00000121410                -0.019687        -0.029221  ...   
10000_ZBTB46_P1P2_ENSG00000130584         0.013731         0.333724  ...   


In [4]:
adata.shape

(11258, 8248)

In [5]:
adata.obs.columns

Index(['UMI_count_unfiltered', 'num_cells_unfiltered', 'num_cells_filtered',
       'control_expr', 'fold_expr', 'pct_expr', 'core_control',
       'mean_leverage_score', 'std_leverage_score', 'energy_test_p_value',
       'anderson_darling_counts', 'mann_whitney_counts', 'z_gemgroup_UMI',
       'mitopercent', 'TE_ratio', 'cnv_score_z'],
      dtype='object')

In [10]:
adata.obs.index

Index(['0_A1BG_P1_ENSG00000121410', '10000_ZBTB46_P1P2_ENSG00000130584',
       '10001_ZBTB47_P1P2_ENSG00000114853',
       '10002_ZBTB48_P1P2_ENSG00000204859',
       '10003_ZBTB49_P1P2_ENSG00000168826', '10004_ZBTB4_P2_ENSG00000174282',
       '10005_ZBTB4_P1_ENSG00000174282', '10006_ZBTB5_P1P2_ENSG00000168795',
       '10007_ZBTB6_P1P2_ENSG00000186130', '10008_ZBTB7A_P1_ENSG00000178951',
       ...
       '9993_ZBTB3_P1P2_ENSG00000185670', '9994_ZBTB40_P1P2_ENSG00000184677',
       '9995_ZBTB41_P1P2_ENSG00000177888', '9996_ZBTB42_P1P2_ENSG00000179627',
       '9997_ZBTB43_P1P2_ENSG00000169155', '9998_ZBTB44_P1P2_ENSG00000196323',
       '9999_ZBTB45_P1P2_ENSG00000119574', '999_C16orf91_P1P2_ENSG00000174109',
       '99_ACSL1_P1P2_ENSG00000151726', '9_AAR2_P1P2_ENSG00000131043'],
      dtype='object', name='gene_transcript', length=11258)

In [6]:
adata.obs.shape

(11258, 16)

In [7]:
adata.obs

UMI_count_unfiltered  num_cells_unfiltered  \
gene_transcript                                                                 
0_A1BG_P1_ENSG00000121410                  11648.256836                   246   
10000_ZBTB46_P1P2_ENSG00000130584          12291.757812                    95   
10001_ZBTB47_P1P2_ENSG00000114853          11295.581055                   318   
10002_ZBTB48_P1P2_ENSG00000204859          12323.345703                    84   
10003_ZBTB49_P1P2_ENSG00000168826          11520.527344                   290   
...                                                 ...                   ...   
9998_ZBTB44_P1P2_ENSG00000196323           10824.015625                   270   
9999_ZBTB45_P1P2_ENSG00000119574           11655.730469                   208   
999_C16orf91_P1P2_ENSG00000174109          11512.895508                   315   
99_ACSL1_P1P2_ENSG00000151726              11877.896484                   415   
9_AAR2_P1P2_ENSG00000131043                10261.093750                    96   

                                   num_cells_filtered  control_expr  \
gene_transcript                                                       
0_A1BG_P1_ENSG00000121410                       237.0      0.171285   
10000_ZBTB46_P1P2_ENSG00000130584                89.0      0.026212   
10001_ZBTB47_P1P2_ENSG00000114853               308.0      0.048150   
10002_ZBTB48_P1P2_ENSG00000204859                83.0      0.134619   
10003_ZBTB49_P1P2_ENSG00000168826               276.0      0.026936   
...                                               ...           ...   
9998_ZBTB44_P1P2_ENSG00000196323                260.0      0.424688   
9999_ZBTB45_P1P2_ENSG00000119574                198.0      0.073588   
999_C16orf91_P1P2_ENSG00000174109               311.0      0.569768   
99_ACSL1_P1P2_ENSG00000151726                   406.0      0.040591   
9_AAR2_P1P2_ENSG00000131043                      93.0      0.174746   

                                   fold_expr  pct_expr  core_control  \
gene_transcript                                                        
0_A1BG_P1_ENSG00000121410           0.023274 -0.976726         False   
10000_ZBTB46_P1P2_ENSG00000130584   0.000000 -1.000000         False   
10001_ZBTB47_P1P2_ENSG00000114853   0.527716 -0.472284         False   
10002_ZBTB48_P1P2_ENSG00000204859   0.163264 -0.836736         False   
10003_ZBTB49_P1P2_ENSG00000168826   0.127473 -0.872527         False   
...                                      ...       ...           ...   
9998_ZBTB44_P1P2_ENSG00000196323    0.137796 -0.862204         False   
9999_ZBTB45_P1P2_ENSG00000119574    0.064148 -0.935852         False   
999_C16orf91_P1P2_ENSG00000174109   0.115690 -0.884310         False   
99_ACSL1_P1P2_ENSG00000151726       0.629563 -0.370437         False   
9_AAR2_P1P2_ENSG00000131043         0.000000 -1.000000         False   

                                   mean_leverage_score  std_leverage_score  \
gene_transcript                                                              
0_A1BG_P1_ENSG00000121410                     0.264768            1.137824   
10000_ZBTB46_P1P2_ENSG00000130584             0.260951            1.095702   
10001_ZBTB47_P1P2_ENSG00000114853             0.277062            1.115502   
10002_ZBTB48_P1P2_ENSG00000204859             0.370545            1.089380   
10003_ZBTB49_P1P2_ENSG00000168826             0.301086            1.099432   
...                                                ...                 ...   
9998_ZBTB44_P1P2_ENSG00000196323              0.331996            1.095451   
9999_ZBTB45_P1P2_ENSG00000119574              0.272367            1.076793   
999_C16orf91_P1P2_ENSG00000174109             0.146729            0.949606   
99_ACSL1_P1P2_ENSG00000151726                 0.148494            1.024573   
9_AAR2_P1P2_ENSG00000131043                   1.423462            1.257388   

                                   energy_test_p_value  \
gene_transcript                                

In [19]:
pertb_genes = adata.obs.index.tolist()
pertb_genes = [gene.split('_')[1] for gene in pertb_genes]
len(pertb_genes)

11258

In [20]:
pertb_genes

['A1BG',
 'ZBTB46',
 'ZBTB47',
 'ZBTB48',
 'ZBTB49',
 'ZBTB4',
 'ZBTB4',
 'ZBTB5',
 'ZBTB6',
 'ZBTB7A',
 'ZBTB7A',
 'C16orf95',
 'ZBTB7B',
 'ZBTB7C',
 'ZBTB7C',
 'ZBTB8A',
 'ZBTB8B',
 'ZBTB8OS',
 'ZBTB9',
 'ZC2HC1C',
 'ZC3H11A',
 'ZC3H12A',
 'C17orf49',
 'ZC3H13',
 'ZC3H14',
 'ZC3H15',
 'ZC3H18',
 'ZC3H3',
 'ZC3H4',
 'ZC3H4',
 'ZC3H7A',
 'ZC3H7B',
 'ZC3H8',
 'C17orf49',
 'ZC3HAV1',
 'ZC3HC1',
 'ZCCHC10',
 'TUT4',
 'ZCCHC17',
 'ZCCHC3',
 'ZCCHC4',
 'TUT7',
 'ZCCHC7',
 'ZCCHC8',
 'C17orf53',
 'ZCCHC9',
 'ZCRB1',
 'ZDHHC12',
 'ZDHHC14',
 'ZDHHC16',
 'ZDHHC17',
 'ZDHHC18',
 'ZDHHC19',
 'ZDHHC20',
 'ZDHHC24',
 'C17orf58',
 'ZDHHC3',
 'ZDHHC4',
 'ZDHHC5',
 'ZDHHC6',
 'ZDHHC7',
 'ZDHHC8',
 'ZDHHC9',
 'ZEB1',
 'CYBC1',
 'ZEB1',
 'ZEB2',
 'ZEB2',
 'ZER1',
 'ZFAND1',
 'ZFAND1',
 'ZFAND2A',
 'ZFAND2B',
 'ZFAND3',
 'ZFAND5',
 'C17orf75',
 'ZFAND6',
 'ZFAT',
 'ZFAT',
 'ZFC3H1',
 'ZFHX2',
 'ZFHX2',
 'ZFHX3',
 'ZFHX4',
 'ZFP14',
 'ZFP1',
 'C17orf80',
 'ZFP28',
 'ZFP2',
 'ZFP30',
 'ZFP36L1',
 'ZFP36L1

In [68]:
unique_genes, idx,  counts = np.unique(pertb_genes, return_index=True, return_counts=True)

In [69]:
unique_genes, counts, idx

(array(['A1BG', 'AAAS', 'AACS', ..., 'ZYX', 'ZZEF1', 'non-targeting'],
       shape=(9867,), dtype='<U14'),
 array([  2,   1,   1, ...,   1,   1, 585], shape=(9867,)),
 array([   0, 3528, 4626, ...,  821,  822,  824], shape=(9867,)))

In [46]:
multiple_perturbed_genes = unique_genes[counts > 1]
for id, count in zip(idx, counts):
    if count > 1:
        print(f"Gene: {pertb_genes[id]}, Count: {count}, for idx: {id}")

Gene: A1BG, Count: 2, for idx: 0
Gene: AARSD1, Count: 2, for idx: 1547
Gene: ABCA7, Count: 2, for idx: 2205
Gene: ABCB9, Count: 2, for idx: 2982
Gene: ABCF1, Count: 2, for idx: 3968
Gene: ABHD14B, Count: 2, for idx: 5068
Gene: ABL2, Count: 2, for idx: 6509
Gene: ABO, Count: 2, for idx: 6729
Gene: ACIN1, Count: 2, for idx: 9492
Gene: ACP5, Count: 2, for idx: 10928
Gene: ACYP1, Count: 2, for idx: 1580
Gene: ADAR, Count: 2, for idx: 1731
Gene: ADAT3, Count: 2, for idx: 1776
Gene: ADK, Count: 2, for idx: 1952
Gene: AGBL5, Count: 2, for idx: 2238
Gene: AGPAT1, Count: 2, for idx: 2349
Gene: AHCYL1, Count: 2, for idx: 2461
Gene: AHNAK, Count: 2, for idx: 2515
Gene: AKNA, Count: 2, for idx: 2859
Gene: AKR1C2, Count: 2, for idx: 2904
Gene: AMPD2, Count: 2, for idx: 3504
Gene: AP2M1, Count: 2, for idx: 4376
Gene: APC, Count: 2, for idx: 4593
Gene: APOBR, Count: 2, for idx: 4781
Gene: APTX, Count: 2, for idx: 4914
Gene: AQP3, Count: 2, for idx: 4947
Gene: ARAP1, Count: 2, for idx: 5001
Gene: ARF5

In [70]:
# arrange counts in descending order
sorted_counts = sorted(counts, reverse=True)

In [73]:
sorted_counts

[np.int64(585),
 np.int64(5),
 np.int64(5),
 np.int64(4),
 np.int64(4),
 np.int64(3),
 np.int64(3),
 np.int64(3),
 np.int64(3),
 np.int64(3),
 np.int64(3),
 np.int64(3),
 np.int64(3),
 np.int64(3),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.int64(2),
 np.

In [59]:
'non-targeting' in unique_genes, 'Control' in unique_genes

(True, False)

In [51]:
len(multiple_perturbed_genes)   # 789 genes are perturbed more than once

789

In [52]:
single_perturbed_genes = unique_genes[counts == 1]
len(single_perturbed_genes)    # 9087 genes are perturbed only once

9078

In [100]:
single_perturbed_genes

array(['AAAS', 'AACS', 'AAGAB', ..., 'ZYG11B', 'ZYX', 'ZZEF1'],
      shape=(9078,), dtype='<U14')

In [18]:
adata.var

gene_name        mean        std        cv  in_matrix  \
gene_id                                                                   
ENSG00000237491   LINC01409    0.114359   0.033798  0.295547       True   
ENSG00000228794   LINC01128    0.178166   0.040590  0.227822       True   
ENSG00000188976       NOC2L    1.375142   0.162536  0.118196       True   
ENSG00000187961      KLHL17    0.102961   0.030863  0.299751       True   
ENSG00000188290        HES4    0.237189   0.057237  0.241315       True   
...                     ...         ...        ...       ...        ...   
ENSG00000198727      MT-CYB  113.862961  11.825139  0.103854       True   
ENSG00000278704  BX004987.1    0.190622   0.042857  0.224826       True   
ENSG00000278384  AL354822.1    0.187205   0.042461  0.226814       True   
ENSG00000271254  AC240274.1    0.158955   0.041703  0.262358       True   
ENSG00000276345  AC004556.3    0.285723   0.056669  0.198334       True   

                     gini  clean_mean  clean_std  clean_cv  
gene_id                                                     
ENSG00000237491  0.153642    0.117205   0.034833  0.297200  
ENSG00000228794  0.122999    0.182778   0.041933  0.229419  
ENSG00000188976  0.063150    1.411069   0.164344  0.116468  
ENSG00000187961  0.162265    0.105230   0.031627  0.300552  
ENSG00000188290  0.127442    0.242638   0.058972  0.243044  
...                   ...         ...        ...       ...  
ENSG00000198727  0.049504  116.198235  11.865300  0.102113  
ENSG00000278704  0.119867    0.195459   0.044112  0.225684  
ENSG00000278384  0.121514    0.191764   0.043733  0.228057  
ENSG00000271254  0.141930    0.163134   0.042716  0.261849  
ENSG00000276345  0.107136    0.293597   0.057869  0.197103  

[8248 rows x 9 columns]

In [108]:
var_gene_id = list(adata.var.index)
print(f'Number of gene_ids in the adata.var: {len(var_gene_id)}')
print(f'Number of unique gene_ids {np.unique(adata.var.index).shape[0]}')

Number of gene_ids in the adata.var: 8248
Number of unique gene_ids 8248


In [109]:
var_gene_name = list(adata.var['gene_name'])
print(f"Number of gene_name in the adata.var['gene_name]: {len(var_gene_name)}")
print(f"Number of unique gene_names: {len(set(var_gene_name))}")

Number of gene_name in the adata.var['gene_name]: 8248
Number of unique gene_names: 8246


In [114]:
var_gene_name.sort()

In [112]:
adata.shape

(11258, 8248)

### It seems the gene_name in the variables are the columns (or variables or features) in the gene expression array

In [ ]:
targeted_genes_only = adata.obs

#### There are a total of 11,258 detected genetic perturbations (expected sgRNA pairs) with a mean coverage 183 cells per perturbation and a median coverage of 171 cells per perturbation after filtering.

### There are 585 non targeting perturbations present in the dataset.

### There are a total of 789 pertubations that appear multiple times and 9078 pertubrations that appear once.

In [75]:
## Why are there multiple entries for the same gene perturbation? 
## This is bulk data so it captures the same perturbation across multiple samples / batches.
multiple_perturbed_genes[:5]

array(['A1BG', 'AARSD1', 'ABCA7', 'ABCB9', 'ABCF1'], dtype='<U14')

In [ ]:
# lets analyse the multiple perturbed genes
multiple_gene_occurence = []
for gene in multiple_perturbed_genes[:5]:
    ids = np.where(adata.obs.index.str.contains(f"_{gene}_") == True)
    occurences = adata.obs.index[ids].tolist()
    multiple_gene_occurence.extend(occurences)
    

In [99]:
multiple_gene_occurence

['0_A1BG_P1_ENSG00000121410',
 '1_A1BG_P2_ENSG00000121410',
 '11_AARSD1_P2_ENSG00000266967',
 '12_AARSD1_P1_ENSG00000266967',
 '17_ABCA7_P2_ENSG00000064687',
 '18_ABCA7_P1_ENSG00000064687',
 '24_ABCB9_P1_ENSG00000150967',
 '25_ABCB9_P2_ENSG00000150967',
 '33_ABCF1_P1_ENSG00000204574',
 '34_ABCF1_P2_ENSG00000204574']

In [ ]:
var_gene_name

['A1BG',
 'AAAS',
 'AACS',
 'AAGAB',
 'AAK1',
 'AAMDC',
 'AAMP',
 'AAR2',
 'AARS',
 'AARS2',
 'AARSD1',
 'AASDH',
 'AASDHPPT',
 'AATF',
 'ABCA7',
 'ABCB10',
 'ABCB6',
 'ABCB7',
 'ABCB8',
 'ABCC1',
 'ABCC10',
 'ABCC4',
 'ABCC5',
 'ABCD3',
 'ABCD4',
 'ABCE1',
 'ABCF1',
 'ABCF2',
 'ABCF3',
 'ABHD10',
 'ABHD11',
 'ABHD12',
 'ABHD13',
 'ABHD14A',
 'ABHD14B',
 'ABHD16A',
 'ABHD17A',
 'ABHD17B',
 'ABHD18',
 'ABHD2',
 'ABHD3',
 'ABHD5',
 'ABI1',
 'ABI2',
 'ABITRAM',
 'ABL1',
 'ABL2',
 'ABO',
 'ABR',
 'ABRACL',
 'ABRAXAS1',
 'ABRAXAS2',
 'ABT1',
 'AC002467.1',
 'AC002558.3',
 'AC003093.1',
 'AC004148.1',
 'AC004556.3',
 'AC004687.1',
 'AC004951.1',
 'AC005261.1',
 'AC005670.3',
 'AC006141.1',
 'AC006329.1',
 'AC006333.2',
 'AC006449.6',
 'AC006504.5',
 'AC007405.1',
 'AC007405.3',
 'AC008124.1',
 'AC008267.5',
 'AC008443.1',
 'AC008555.4',
 'AC008875.3',
 'AC009403.1',
 'AC010642.2',
 'AC012146.1',
 'AC012360.3',
 'AC012615.1',
 'AC015574.1',
 'AC015813.1',
 'AC016065.1',
 'AC016074.2',
 'AC016

In [121]:
unique_genes.shape

(9867,)

In [119]:
print(unique_genes)

['A1BG' 'AAAS' 'AACS' ... 'ZYX' 'ZZEF1' 'non-targeting']


### Clustered Mean Gene expression Figure 2-4 in the paper

In [5]:
cluster_gene_df = pd.read_csv('data/clustered_mean_gene_expression_figs2-4.csv')
cluster_gene_df

/tmp/ipykernel_24300/1940324100.py:1: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.
  cluster_gene_df = pd.read_csv('data/clustered_mean_gene_expression_figs2-4.csv')


gene_transcript Unnamed: 1  10005_ZBTB4_P1_ENSG00000174282  \
0            cluster        NaN                       19.000000   
1          gene_name    cluster                             NaN   
2            CYP51A1          4                        0.015538   
3                BAD         -1                        0.075532   
4               CD99         -1                       -0.185194   
...              ...        ...                             ...   
2316      AC139493.2         -1                       -0.422495   
2317         POLR2J3         -1                        0.379631   
2318      BX890604.2         -1                        0.060350   
2319      AC016074.2          5                        0.364247   
2320      AP001531.1         15                        0.224872   

      10006_ZBTB5_P1P2_ENSG00000168795  10015_ZBTB8OS_P1P2_ENSG00000176261  \
0                            -1.000000                           -1.000000   
1                                  NaN                                 NaN   
2                            -0.169297                           -0.141619   
3                            -0.039656                           -0.133556   
4                             0.067921                            0.138278   
...                                ...                                 ...   
2316                         -0.064950                           -0.086410   
2317                          0.000678                            0.084222   
2318                          0.015746                            0.098428   
2319                         -0.016822                            0.285028   
2320                          0.040849                            0.071846   

      10020_ZC3H13_P1P2_ENSG00000123200  10023_ZC3H18_P1P2_ENSG00000158545  \
0                              4.000000                          -1.000000   
1                                   NaN                                NaN   
2                             -0.068998                           0.019589   
3                              0.208468                           0.355862   
4                              0.081160                          -0.203804   
...                                 ...                                ...   
2316                          -0.023713                          -0.344314   
2317                          -0.112965                           0.352633   
2318                          -0.192752                           0.080785   
2319                          -0.185722                          -0.163307   
2320                           0.027330                          -0.029889   

      10024_ZC3H3_P1P2_ENSG00000014164  10025_ZC3H4_P1_ENSG00000130749  \
0                             8.000000                       -1.000000   
1                                  NaN                             NaN   
2                            -0.017413                       -0.341628   
3                             0.134834                       -0.133954   
4                             0.125460                       -0.038213   
...                                ...                             ...   
2316                         -0.050482                       -0.082852   
2317                         -0.123764                        0.408237   
2318                          0.083702                       -0.067631   
2319                          0.036043                        0.028476   
2320                         -0.033591                        0.133265   

      10039_ZCCHC8_P1P2_ENSG00000033030  ...  \
0                             -1.000000  ...   
1                                   NaN  ...   
2                             -0.054714  ...   
3                              0.031147  ...   
4                             -0.051570  ...   
...                                 ...  ...   
2316                          -0.001369  ...   
2317                          -0.019059  ...   
2318 

In [69]:
cluster_gene_df.columns

Index(['gene_transcript', 'Unnamed: 1', '10005_ZBTB4_P1_ENSG00000174282',
       '10006_ZBTB5_P1P2_ENSG00000168795',
       '10015_ZBTB8OS_P1P2_ENSG00000176261',
       '10020_ZC3H13_P1P2_ENSG00000123200',
       '10023_ZC3H18_P1P2_ENSG00000158545', '10024_ZC3H3_P1P2_ENSG00000014164',
       '10025_ZC3H4_P1_ENSG00000130749', '10039_ZCCHC8_P1P2_ENSG00000033030',
       ...
       '9925_YEATS2_P1P2_ENSG00000163872', '9926_YEATS4_P1P2_ENSG00000127337',
       '9940_YME1L1_P1P2_ENSG00000136758', '995_C16orf72_P1_ENSG00000182831',
       '9971_ZBTB14_P1P2_ENSG00000198081', '9975_ZBTB17_P1P2_ENSG00000116809',
       '998_C16orf87_P1P2_ENSG00000155330', '9992_ZBTB39_P1P2_ENSG00000166860',
       '9998_ZBTB44_P1P2_ENSG00000196323', '9_AAR2_P1P2_ENSG00000131043'],
      dtype='object', length=1975)

In [70]:
strong_perturbation_genes = cluster_gene_df.columns.tolist()[2:]
print(f'There are in total {len(strong_perturbation_genes)} strong pertubations')

There are in total 1973 strong pertubations


In [71]:
# extract the gene names along with P1 or P2 from the strong perturbation genes:
strong_perturbation_genes_names = [i.split('_')[1] for i in strong_perturbation_genes]
strong_perturbation_genes_promoter = [i.split('_')[2] for i in strong_perturbation_genes]
strong_perturbation_genes_code = [i.split('_')[3] for i in strong_perturbation_genes]

In [72]:
len(strong_perturbation_genes_names)

1973

In [74]:
len(set(strong_perturbation_genes_names))

1969

There are certain repetition of genes if we only consider the gene name but if we consider gene name along with the promoter then there are no repetition, that is there are 1973 unique strong perturbations!

This means some genes are perturbed more than once but with different Promoter!!


In [75]:
strong_perturbation_genes_name_promoter = [f"{name}_{promoter}" for name, promoter in zip(strong_perturbation_genes_names, strong_perturbation_genes_promoter)]
strong_perturbation_genes_name_promoter[:10]

['ZBTB4_P1',
 'ZBTB5_P1P2',
 'ZBTB8OS_P1P2',
 'ZC3H13_P1P2',
 'ZC3H18_P1P2',
 'ZC3H3_P1P2',
 'ZC3H4_P1',
 'ZCCHC8_P1P2',
 'ZCCHC9_P1P2',
 'ZCRB1_P1P2']

In [76]:
len(strong_perturbation_genes_name_promoter)

1973

In [77]:
# lets check the perturbation cluster of these genes
prtb_cluster = np.array(cluster_gene_df.iloc[0][2:], dtype=int)
prtb_cluster

array([19, -1, -1, ..., -1, -1, 57], shape=(1973,))

In [78]:
np.unique(prtb_cluster, return_counts=True)

(array([-1,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15,
        16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32,
        33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49,
        50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63]),
 array([1268,    4,    6,    8,    5,    6,   10,   15,    8,   20,    5,
           4,    4,    5,    8,    6,    6,    4,    7,    7,    4,    4,
           4,    5,    5,    5,    4,    4,    6,    4,    7,    5,   35,
           4,   10,    5,    7,    6,   15,    4,    4,    6,    5,    6,
          10,   10,    9,    4,    4,    9,    5,   19,    4,   26,   23,
           5,    4,   11,   97,   43,   10,   13,   53,   40,    4]))

There are 64 perturbation clusters as mentioned in the paper, -1 indicates no cluster was assigned to the gene!

In [79]:
# Lets get the list of 2319 highly variable genes as mentioned in the paper...
responsive_genes = cluster_gene_df['gene_transcript'].to_list()[2:]
print(f'There are in total {len(responsive_genes)} responsive genes!')

There are in total 2319 responsive genes!


In [81]:
# lets save the lists of highly variable genes and strongly perturbation genes
strong_prtb_df = pd.DataFrame({'gene_name': strong_perturbation_genes_names,
                               'gene_name_promoter': strong_perturbation_genes_name_promoter,
                               'gene_code': strong_perturbation_genes_code,
                               'gene_promoter': strong_perturbation_genes_promoter,
                               'cluster': prtb_cluster})
strong_prtb_df

gene_name gene_name_promoter        gene_code gene_promoter  cluster
0        ZBTB4           ZBTB4_P1  ENSG00000174282            P1       19
1        ZBTB5         ZBTB5_P1P2  ENSG00000168795          P1P2       -1
2      ZBTB8OS       ZBTB8OS_P1P2  ENSG00000176261          P1P2       -1
3       ZC3H13        ZC3H13_P1P2  ENSG00000123200          P1P2        4
4       ZC3H18        ZC3H18_P1P2  ENSG00000158545          P1P2       -1
...        ...                ...              ...           ...      ...
1968    ZBTB17        ZBTB17_P1P2  ENSG00000116809          P1P2        7
1969  C16orf87      C16orf87_P1P2  ENSG00000155330          P1P2        2
1970    ZBTB39        ZBTB39_P1P2  ENSG00000166860          P1P2       -1
1971    ZBTB44        ZBTB44_P1P2  ENSG00000196323          P1P2       -1
1972      AAR2          AAR2_P1P2  ENSG00000131043          P1P2       57

[1973 rows x 5 columns]

In [82]:
prtb_genes_not_assigned_to_any_cluster = strong_prtb_df[strong_prtb_df['cluster']==-1]

In [83]:
len(prtb_genes_not_assigned_to_any_cluster)

1268

In [85]:
clusters, gene_per_cluster = np.unique(strong_prtb_df["cluster"].to_list(), return_counts=True)
for cls, number in zip(clusters, gene_per_cluster):
    print(f'For Perturbation cluster number {cls}, there are {number} number of genes assigned to it!')

For Perturbation cluster number -1, there are 1268 number of genes assigned to it!
For Perturbation cluster number 0, there are 4 number of genes assigned to it!
For Perturbation cluster number 1, there are 6 number of genes assigned to it!
For Perturbation cluster number 2, there are 8 number of genes assigned to it!
For Perturbation cluster number 3, there are 5 number of genes assigned to it!
For Perturbation cluster number 4, there are 6 number of genes assigned to it!
For Perturbation cluster number 5, there are 10 number of genes assigned to it!
For Perturbation cluster number 6, there are 15 number of genes assigned to it!
For Perturbation cluster number 7, there are 8 number of genes assigned to it!
For Perturbation cluster number 8, there are 20 number of genes assigned to it!
For Perturbation cluster number 9, there are 5 number of genes assigned to it!
For Perturbation cluster number 10, there are 4 number of genes assigned to it!
For Perturbation cluster number 11, there ar

There are 1286 genes in the Perturbation cluster that was not assigned to any cluster!

In [87]:
gene_expression_cluster = cluster_gene_df['Unnamed: 1'].to_list()[2:]
print(f'There are total {len(gene_expression_cluster)} gene expression clusters')

There are total 2319 gene expression clusters


In [89]:
gene_clusters, gene_per_gene_cluster = np.unique(gene_expression_cluster, return_counts=True)
for cls, number in zip(gene_clusters, gene_per_gene_cluster):
    print(f'For Gene Expression cluster number {cls}, there are {number} number of genes assigned to it!')

For Gene Expression cluster number -1, there are 1133 number of genes assigned to it!
For Gene Expression cluster number 0, there are 15 number of genes assigned to it!
For Gene Expression cluster number 1, there are 12 number of genes assigned to it!
For Gene Expression cluster number 10, there are 18 number of genes assigned to it!
For Gene Expression cluster number 11, there are 23 number of genes assigned to it!
For Gene Expression cluster number 12, there are 22 number of genes assigned to it!
For Gene Expression cluster number 13, there are 22 number of genes assigned to it!
For Gene Expression cluster number 14, there are 48 number of genes assigned to it!
For Gene Expression cluster number 15, there are 37 number of genes assigned to it!
For Gene Expression cluster number 16, there are 14 number of genes assigned to it!
For Gene Expression cluster number 17, there are 72 number of genes assigned to it!
For Gene Expression cluster number 18, there are 52 number of genes assigned

There are 1133 genes (out of 2319 genes) in the Gene Expression Cluster, that was not assigned to any cluster!

In [195]:
highly_responsive_gene_df = pd.DataFrame({'gene_name':responsive_genes,
                                          'cluster': gene_expression_cluster})
highly_responsive_gene_df

gene_name cluster
0        CYP51A1       4
1            BAD      -1
2           CD99      -1
3         MAD1L1      29
4          CFLAR      19
...          ...     ...
2314  AC139493.2      -1
2315     POLR2J3      -1
2316  BX890604.2      -1
2317  AC016074.2       5
2318  AP001531.1      15

[2319 rows x 2 columns]

In [197]:
#highly_responsive_gene_df.to_csv('data/highly_responsive_gene_2319.csv', index=0)

In [ ]:
#strong_prtb_df.to_csv('data/strong_perturbations_1973.csv', index=0)

In [200]:
adata.shape

(11258, 8248)

### Reduce the andata to highly responsive genes and strong perturbations

In [22]:
# read the df for strong perturbation genes and highly variable genes for which the clusters were formed:

SPG_df = pd.read_csv('data/strong_perturbations_1973.csv')
HVG_df = pd.read_csv('data/highly_responsive_gene_2319.csv')

In [23]:
SPG_df.head()

gene_name gene_name_promoter        gene_code gene_promoter  cluster
0     ZBTB4           ZBTB4_P1  ENSG00000174282            P1       19
1     ZBTB5         ZBTB5_P1P2  ENSG00000168795          P1P2       -1
2   ZBTB8OS       ZBTB8OS_P1P2  ENSG00000176261          P1P2       -1
3    ZC3H13        ZC3H13_P1P2  ENSG00000123200          P1P2        4
4    ZC3H18        ZC3H18_P1P2  ENSG00000158545          P1P2       -1

In [24]:
SPG_df.shape

(1973, 5)

In [25]:
HVG_df.shape

(2319, 2)

In [345]:
HVG_df.head()

gene_name  cluster
0   CYP51A1        4
1       BAD       -1
2      CD99       -1
3    MAD1L1       29
4     CFLAR       19

In [27]:
SPG_df['gene_name'].nunique(), HVG_df['gene_name'].nunique()

(1969, 2319)

In [29]:
# lets create an intersection and union of these genes:
SPG_union_HVG = set(SPG_df['gene_name'].to_list()) | set(HVG_df['gene_name'].to_list())
print(f'There are {len(SPG_df)} genes for strong pertubation and {len(HVG_df)} highly variable genes.')
print(f'The union of both has {len(SPG_union_HVG)} genes')


There are 1973 genes for strong pertubation and 2319 highly variable genes.
The union of both has 3606 genes


In [28]:
SPG_int_HVG = set(SPG_df['gene_name'].to_list()) & set(HVG_df['gene_name'].to_list())
print(f'The intersection of both has {len(SPG_int_HVG)} genes.')

The intersection of both has 682 genes.


In [30]:
expression_df = adata.to_df()
expression_df

gene_id                            ENSG00000237491  ENSG00000228794  \
gene_transcript                                                       
0_A1BG_P1_ENSG00000121410                 0.028642         0.046873   
10000_ZBTB46_P1P2_ENSG00000130584         0.055915        -0.058623   
10001_ZBTB47_P1P2_ENSG00000114853         0.030780         0.038029   
10002_ZBTB48_P1P2_ENSG00000204859        -0.104988        -0.168258   
10003_ZBTB49_P1P2_ENSG00000168826         0.140855        -0.035711   
...                                            ...              ...   
9998_ZBTB44_P1P2_ENSG00000196323          0.017919         0.013294   
9999_ZBTB45_P1P2_ENSG00000119574          0.004931         0.061982   
999_C16orf91_P1P2_ENSG00000174109         0.008794         0.116086   
99_ACSL1_P1P2_ENSG00000151726            -0.017405        -0.050810   
9_AAR2_P1P2_ENSG00000131043              -0.015396         0.017742   

gene_id                            ENSG00000188976  ENSG00000187961  \
gene_transcript                                                       
0_A1BG_P1_ENSG00000121410                -0.007014        -0.043553   
10000_ZBTB46_P1P2_ENSG00000130584         0.106205         0.025113   
10001_ZBTB47_P1P2_ENSG00000114853         0.067629        -0.136385   
10002_ZBTB48_P1P2_ENSG00000204859        -0.044452        -0.057588   
10003_ZBTB49_P1P2_ENSG00000168826        -0.089407         0.096020   
...                                            ...              ...   
9998_ZBTB44_P1P2_ENSG00000196323          0.166343         0.202679   
9999_ZBTB45_P1P2_ENSG00000119574          0.024869         0.019066   
999_C16orf91_P1P2_ENSG00000174109        -0.067448        -0.006761   
99_ACSL1_P1P2_ENSG00000151726             0.012079        -0.077181   
9_AAR2_P1P2_ENSG00000131043               0.313340         0.215871   

gene_id                            ENSG00000188290  ENSG00000187608  \
gene_transcript                                                       
0_A1BG_P1_ENSG00000121410                 0.049834         0.055918   
10000_ZBTB46_P1P2_ENSG00000130584         0.272732        -0.114054   
10001_ZBTB47_P1P2_ENSG00000114853        -0.035681        -0.055437   
10002_ZBTB48_P1P2_ENSG00000204859        -0.024554        -0.002767   
10003_ZBTB49_P1P2_ENSG00000168826         0.084299         0.033209   
...                                            ...              ...   
9998_ZBTB44_P1P2_ENSG00000196323          0.026368        -0.026491   
9999_ZBTB45_P1P2_ENSG00000119574         -0.035541        -0.036628   
999_C16orf91_P1P2_ENSG00000174109         0.070513        -0.057997   
99_ACSL1_P1P2_ENSG00000151726            -0.011902        -0.069777   
9_AAR2_P1P2_ENSG00000131043              -0.090744         0.256190   

gene_id                            ENSG00000078808  ENSG00000176022  \
gene_transcript                                                       
0_A1BG_P1_ENSG00000121410                 0.056130        -0.009680   
10000_ZBTB46_P1P2_ENSG00000130584         0.174116        -0.106970   
10001_ZBTB47_P1P2_ENSG00000114853         0.027704        -0.056662   
10002_ZBTB48_P1P2_ENSG00000204859        -0.034466         0.020218   
10003_ZBTB49_P1P2_ENSG00000168826        -0.039355        -0.126822   
...                                            ...              ...   
9998_ZBTB44_P1P2_ENSG00000196323          0.081814         0.051754   
9999_ZBTB45_P1P2_ENSG00000119574          0.021431        -0.088374   
999_C16orf91_P1P2_ENSG00000174109         0.027870        -0.021607   
99_ACSL1_P1P2_ENSG00000151726             0.077219        -0.024646   
9_AAR2_P1P2_ENSG00000131043               0.116805         0.099588   

gene_id                            ENSG00000160087  ENSG00000131584  ...  \
gene_transcript                                                      ...   
0_A1BG_P1_ENSG00000121410                -0.019687        -0.029221  ...   
10000_ZBTB46_P1P2_ENSG00000130584         0.013731         0.333724  ...   


In [31]:
sum(expression_df.index.str.contains('NON-TARGETING', case=False))

np.int64(585)

The rows are the pertubed gene names and the columns are the features (or all gene_name)

In [32]:
gene_exp_prtb_genes = expression_df.index.to_list()
len(gene_exp_prtb_genes)

11258

In [33]:
gene_exp_vars = expression_df.columns.to_list()
len(gene_exp_vars)

8248

In [21]:
adata.shape

(11258, 8248)

In [11]:
expression_df.shape

(11258, 8248)

In [23]:
len(gene_exp_vars), np.unique(gene_exp_vars).shape[0]

(8248, 8248)

In [24]:
adata.var['gene_name']

gene_id
ENSG00000237491     LINC01409
ENSG00000228794     LINC01128
ENSG00000188976         NOC2L
ENSG00000187961        KLHL17
ENSG00000188290          HES4
                      ...    
ENSG00000198727        MT-CYB
ENSG00000278704    BX004987.1
ENSG00000278384    AL354822.1
ENSG00000271254    AC240274.1
ENSG00000276345    AC004556.3
Name: gene_name, Length: 8248, dtype: category
Categories (8246, object): ['A1BG', 'AAAS', 'AACS', 'AAGAB', ..., 'ZYG11A', 'ZYG11B', 'ZYX', 'ZZEF1']

In [35]:
code_to_name_var_dict = adata.var['gene_name'].to_dict()
len(code_to_name_var_dict)

8248

In [36]:
# lets create a union of gene code which will be used for various purposes:
code_from_prtrb = [i.split('_')[-1] for i in gene_exp_prtb_genes]
total_gene_code = set(gene_exp_vars) | set(code_from_prtrb)
print(f'There are total {len(total_gene_code)} gene code from the adata (expression data) after combining the variables(or features) and the pertubations (or rows)')

There are total 10419 gene code from the adata (expression data) after combining the variables(or features) and the pertubations (or rows)


In [37]:
len(code_from_prtrb), np.unique(code_from_prtrb).shape[0]

(11258, 9854)

In [38]:
# lets create a union of gene name from the perturbation and the variables in the gene expression dataframe:
gene_name_from_prtrb = [i.split('_')[1] for i in gene_exp_prtb_genes]
total_gene_name_from_exp_matrix  = set(gene_name_from_prtrb) | set(code_to_name_var_dict.values())
print(f'Union of gene names in the expression matrix is {len(total_gene_name_from_exp_matrix)}')

Union of gene names in the expression matrix is 10429


In [39]:
adata_variables = adata.var.index.to_list()
# this should be strictly equal to the columns of the gene expression dataframe!
print(f' Are the columns of the gene expression df same as the variables in the adata ? \n {adata_variables == gene_exp_vars}')

 Are the columns of the gene expression df same as the variables in the adata ? 
 True


#### this contains non targeting perturbations (585) as well and pertubations with different promoters.
#### Thats why there is a repetition of gene (or gene code)
#### All variables (or features) are unique (w/o repetition) in the expression matrix but the perturbations are repetitive.

In [40]:
len(SPG_union_HVG)

3606

In [41]:
indices_to_keep = []
for idx in gene_exp_prtb_genes:
    name = idx.split('_')[1]
    if name in SPG_union_HVG:
        indices_to_keep.append(idx)

print(len(indices_to_keep))

3758


In [42]:
reduced_df = expression_df.loc[indices_to_keep].copy(deep=True)

In [43]:
reduced_df.shape

(3758, 8248)

In [44]:
len(set(reduced_df.index))

3758

In [45]:
reduced_df

gene_id                             ENSG00000237491  ENSG00000228794  \
gene_transcript                                                        
10004_ZBTB4_P2_ENSG00000174282             0.018625         0.065188   
10005_ZBTB4_P1_ENSG00000174282             0.141025        -0.043322   
10006_ZBTB5_P1P2_ENSG00000168795           0.042970         0.250034   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.154239        -0.147156   
10020_ZC3H13_P1P2_ENSG00000123200          0.087844         0.467162   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311             0.036313        -0.093249   
9991_ZBTB38_P1_ENSG00000177311            -0.012903        -0.035431   
9992_ZBTB39_P1P2_ENSG00000166860          -0.095441        -0.005933   
9998_ZBTB44_P1P2_ENSG00000196323           0.017919         0.013294   
9_AAR2_P1P2_ENSG00000131043               -0.015396         0.017742   

gene_id                             ENSG00000188976  ENSG00000187961  \
gene_transcript                                                        
10004_ZBTB4_P2_ENSG00000174282             0.116162         0.117416   
10005_ZBTB4_P1_ENSG00000174282            -0.030201        -0.088615   
10006_ZBTB5_P1P2_ENSG00000168795          -0.094131         0.093585   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.042972        -0.122381   
10020_ZC3H13_P1P2_ENSG00000123200         -0.120841        -0.013023   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.023088         0.074530   
9991_ZBTB38_P1_ENSG00000177311             0.032109         0.040046   
9992_ZBTB39_P1P2_ENSG00000166860          -0.054780        -0.051475   
9998_ZBTB44_P1P2_ENSG00000196323           0.166343         0.202679   
9_AAR2_P1P2_ENSG00000131043                0.313340         0.215871   

gene_id                             ENSG00000188290  ENSG00000187608  \
gene_transcript                                                        
10004_ZBTB4_P2_ENSG00000174282             0.016931        -0.000809   
10005_ZBTB4_P1_ENSG00000174282             0.332238         0.097081   
10006_ZBTB5_P1P2_ENSG00000168795          -0.013079         0.041780   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.087034         0.015089   
10020_ZC3H13_P1P2_ENSG00000123200          0.511434         0.690666   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.017826        -0.028234   
9991_ZBTB38_P1_ENSG00000177311            -0.043616         0.105273   
9992_ZBTB39_P1P2_ENSG00000166860          -0.106928         0.206940   
9998_ZBTB44_P1P2_ENSG00000196323           0.026368        -0.026491   
9_AAR2_P1P2_ENSG00000131043               -0.090744         0.256190   

gene_id                             ENSG00000078808  ENSG00000176022  \
gene_transcript                                                        
10004_ZBTB4_P2_ENSG00000174282            -0.061713         0.082389   
10005_ZBTB4_P1_ENSG00000174282            -0.142072        -0.111593   
10006_ZBTB5_P1P2_ENSG00000168795          -0.054414         0.025386   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.154403         0.168626   
10020_ZC3H13_P1P2_ENSG00000123200          0.775491        -0.006288   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.021374        -0.072778   
9991_ZBTB38_P1_ENSG00000177311             0.026275        -0.063379   
9992_ZBTB39_P1P2_ENSG00000166860           0.194758         0.041263   
9998_ZBTB44_P1P2_ENSG00000196323           0.081814         0.051754   
9_AAR2_P1P2_ENSG00000131043                0.116805         0.099588   

gene_id                             ENSG00000160087  ENSG00000131584  ...  \
gene_transcript                                                       ...   
10004_ZBTB4_P2_ENSG00000174282            -0.023648         0.032292  ...   
10005_ZBTB4_P1_ENSG00

In [46]:
len(reduced_df.columns)

8248

In [47]:
set(reduced_df.columns) == set(gene_exp_vars)

True

In [48]:
len(reduced_df.columns), len(gene_exp_vars)

(8248, 8248)

In [ ]:
columns_to_keep = []
for col in gene_exp_vars:
    name = code_to_name_var_dict[col]
    if name in SPG_union_HVG:
        columns_to_keep.append(col)

print(len(columns_to_keep))

3460


In [50]:
reduced_df = reduced_df[columns_to_keep].copy()

In [51]:
reduced_df.shape

(3758, 3460)

In [52]:
reduced_df

gene_id                             ENSG00000188976  ENSG00000187961  \
gene_transcript                                                        
10004_ZBTB4_P2_ENSG00000174282             0.116162         0.117416   
10005_ZBTB4_P1_ENSG00000174282            -0.030201        -0.088615   
10006_ZBTB5_P1P2_ENSG00000168795          -0.094131         0.093585   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.042972        -0.122381   
10020_ZC3H13_P1P2_ENSG00000123200         -0.120841        -0.013023   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.023088         0.074530   
9991_ZBTB38_P1_ENSG00000177311             0.032109         0.040046   
9992_ZBTB39_P1P2_ENSG00000166860          -0.054780        -0.051475   
9998_ZBTB44_P1P2_ENSG00000196323           0.166343         0.202679   
9_AAR2_P1P2_ENSG00000131043                0.313340         0.215871   

gene_id                             ENSG00000187608  ENSG00000160087  \
gene_transcript                                                        
10004_ZBTB4_P2_ENSG00000174282            -0.000809        -0.023648   
10005_ZBTB4_P1_ENSG00000174282             0.097081        -0.050281   
10006_ZBTB5_P1P2_ENSG00000168795           0.041780         0.011111   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.015089        -0.010220   
10020_ZC3H13_P1P2_ENSG00000123200          0.690666         0.369722   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.028234        -0.102966   
9991_ZBTB38_P1_ENSG00000177311             0.105273         0.070274   
9992_ZBTB39_P1P2_ENSG00000166860           0.206940         0.023562   
9998_ZBTB44_P1P2_ENSG00000196323          -0.026491        -0.038248   
9_AAR2_P1P2_ENSG00000131043                0.256190         0.451411   

gene_id                             ENSG00000127054  ENSG00000162576  \
gene_transcript                                                        
10004_ZBTB4_P2_ENSG00000174282            -0.089110         0.064874   
10005_ZBTB4_P1_ENSG00000174282             0.053923         0.047092   
10006_ZBTB5_P1P2_ENSG00000168795           0.049495         0.068986   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.021991        -0.140862   
10020_ZC3H13_P1P2_ENSG00000123200          0.229487         0.635588   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311             0.021527        -0.110965   
9991_ZBTB38_P1_ENSG00000177311             0.090874         0.093788   
9992_ZBTB39_P1P2_ENSG00000166860          -0.088279        -0.061373   
9998_ZBTB44_P1P2_ENSG00000196323           0.146278        -0.032702   
9_AAR2_P1P2_ENSG00000131043               -0.083404         0.057765   

gene_id                             ENSG00000175756  ENSG00000242485  \
gene_transcript                                                        
10004_ZBTB4_P2_ENSG00000174282             0.027806         0.135219   
10005_ZBTB4_P1_ENSG00000174282             0.025221        -0.003499   
10006_ZBTB5_P1P2_ENSG00000168795          -0.062200         0.044283   
10015_ZBTB8OS_P1P2_ENSG00000176261        -0.131819         0.023055   
10020_ZC3H13_P1P2_ENSG00000123200          0.187934         0.145091   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.041697        -0.002252   
9991_ZBTB38_P1_ENSG00000177311            -0.022295         0.051294   
9992_ZBTB39_P1P2_ENSG00000166860           0.207843         0.118381   
9998_ZBTB44_P1P2_ENSG00000196323           0.019084        -0.108255   
9_AAR2_P1P2_ENSG00000131043                0.014528         0.280991   

gene_id                             ENSG00000160072  ENSG00000197785  ...  \
gene_transcript                                                       ...   
10004_ZBTB4_P2_ENSG00000174282            -0.091799         0.126250  ...   
10005_ZBTB4_P1_ENSG00

In [55]:
SPG_but_not_HVG = []
for g in SPG_df['gene_name'].to_list():
    if g not in HVG_df['gene_name'].to_list():
        SPG_but_not_HVG.append(g)

print(len(SPG_but_not_HVG))

1290


In [56]:
HVG_but_not_SPG = []
for g in HVG_df['gene_name'].to_list():
    if g not in SPG_df['gene_name'].to_list():
        HVG_but_not_SPG.append(g)

print(len(HVG_but_not_SPG))

1637


In [57]:
len(SPG_df['gene_name'].to_list()) - len(set(SPG_df['gene_name'].to_list()) & set(HVG_df['gene_name'].to_list()))

1291

Thus all Highly Variable genes are in the column of the expression matrix.

However the number of columns is not equal to 3606!

Probably because some of the Strong Perturbation genes were not highly expressive

In [ ]:
reduced_columns = reduced_df.columns.to_list()  # columns are gene expression (features)
len(reduced_columns)

3460

In [61]:
name_to_code_var_dict = {y: x for x, y in code_to_name_var_dict.items()}

In [72]:
for g in HVG_df['gene_name'].to_list():
    if name_to_code_var_dict[g] in reduced_columns:
        continue
    else:
        print(f'Highly Variable Gene {g} not in reduced expression Dataframe')

All HVG are in the columns of reduced Dataframe!

In [426]:
all(adata.var.index == expression_df.columns)

True

In [74]:
SPG_not_in_expr_column = []
i = 0
for g in SPG_df['gene_code']:
    if g not in reduced_df.columns:
        SPG_not_in_expr_column.append(g)
    else:
        i+=1

print(f'Out of the total {len(SPG_df)} strong perturbation gene, only {i} number of genes are in columns, rest {len(SPG_not_in_expr_column)} are not in columns.')

Out of the total 1973 strong perturbation gene, only 1825 number of genes are in columns, rest 148 are not in columns.


It does not make sense to keep the genes which were perturbed on but they are not in columns, 

since we dont have a behaviour profile for these to use it for performing Gene Expression Clustering!

In [97]:
reduced_gene_name_index = [i.split("_")[1] for i in reduced_df.index.to_list()]
len(reduced_gene_name_index)

3758

In [76]:
HVG_not_in_expr_rows = []
i = 0
for g in HVG_df['gene_name']:
    if g not in reduced_gene_name_index:
        HVG_not_in_expr_rows.append(g)
    else:
        i+=1

print(f'Out of the total {len(HVG_df)} Highly Variable Gene, only {i} number of genes are in rows, rest {len(HVG_not_in_expr_rows)} are not in rows.')

Out of the total 2319 Highly Variable Gene, only 2199 number of genes are in rows, rest 120 are not in rows.


For the case of HVG not being in rows means that we dont have data for their own perturbation, 

but we still have their behaviour profile (i.e they are in columns) for other perturbations, 

so we do not need to remove anything!

In [95]:
reduced_df.shape

(3758, 3460)

In [96]:
print(f'There are {len(indices_to_keep)} indices (or perturbations) retained based on the Union SPG and HVG')
print(f'However there are {len(SPG_union_HVG)} genes when union of SPG and HVG are taken')
print(f'In total {len(indices_to_keep) - len(SPG_union_HVG)} are repeated or there is a mistake in the code!')

There are 3758 indices (or perturbations) retained based on the Union SPG and HVG
However there are 3606 genes when union of SPG and HVG are taken
In total 152 are repeated or there is a mistake in the code!


In [98]:
SPG_not_in_exp_matrix_row = []
for i in SPG_df['gene_name'].to_list():
    if i not in reduced_gene_name_index:
        SPG_not_in_exp_matrix_row.append(i)

In [99]:
len(SPG_not_in_exp_matrix_row)

0

SPG were selected based on the gene name and not code, so there are repetitions of genes

In [101]:
HVG_df[HVG_df.duplicated('gene_name')]

Empty DataFrame
Columns: [gene_name, cluster]
Index: []

In [102]:
SPG_df['gene_name'].nunique()

1969

In [103]:
reduced_df

gene_id                             ENSG00000188976  ENSG00000187961  \
gene_transcript                                                        
10004_ZBTB4_P2_ENSG00000174282             0.116162         0.117416   
10005_ZBTB4_P1_ENSG00000174282            -0.030201        -0.088615   
10006_ZBTB5_P1P2_ENSG00000168795          -0.094131         0.093585   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.042972        -0.122381   
10020_ZC3H13_P1P2_ENSG00000123200         -0.120841        -0.013023   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.023088         0.074530   
9991_ZBTB38_P1_ENSG00000177311             0.032109         0.040046   
9992_ZBTB39_P1P2_ENSG00000166860          -0.054780        -0.051475   
9998_ZBTB44_P1P2_ENSG00000196323           0.166343         0.202679   
9_AAR2_P1P2_ENSG00000131043                0.313340         0.215871   

gene_id                             ENSG00000187608  ENSG00000160087  \
gene_transcript                                                        
10004_ZBTB4_P2_ENSG00000174282            -0.000809        -0.023648   
10005_ZBTB4_P1_ENSG00000174282             0.097081        -0.050281   
10006_ZBTB5_P1P2_ENSG00000168795           0.041780         0.011111   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.015089        -0.010220   
10020_ZC3H13_P1P2_ENSG00000123200          0.690666         0.369722   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.028234        -0.102966   
9991_ZBTB38_P1_ENSG00000177311             0.105273         0.070274   
9992_ZBTB39_P1P2_ENSG00000166860           0.206940         0.023562   
9998_ZBTB44_P1P2_ENSG00000196323          -0.026491        -0.038248   
9_AAR2_P1P2_ENSG00000131043                0.256190         0.451411   

gene_id                             ENSG00000127054  ENSG00000162576  \
gene_transcript                                                        
10004_ZBTB4_P2_ENSG00000174282            -0.089110         0.064874   
10005_ZBTB4_P1_ENSG00000174282             0.053923         0.047092   
10006_ZBTB5_P1P2_ENSG00000168795           0.049495         0.068986   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.021991        -0.140862   
10020_ZC3H13_P1P2_ENSG00000123200          0.229487         0.635588   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311             0.021527        -0.110965   
9991_ZBTB38_P1_ENSG00000177311             0.090874         0.093788   
9992_ZBTB39_P1P2_ENSG00000166860          -0.088279        -0.061373   
9998_ZBTB44_P1P2_ENSG00000196323           0.146278        -0.032702   
9_AAR2_P1P2_ENSG00000131043               -0.083404         0.057765   

gene_id                             ENSG00000175756  ENSG00000242485  \
gene_transcript                                                        
10004_ZBTB4_P2_ENSG00000174282             0.027806         0.135219   
10005_ZBTB4_P1_ENSG00000174282             0.025221        -0.003499   
10006_ZBTB5_P1P2_ENSG00000168795          -0.062200         0.044283   
10015_ZBTB8OS_P1P2_ENSG00000176261        -0.131819         0.023055   
10020_ZC3H13_P1P2_ENSG00000123200          0.187934         0.145091   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.041697        -0.002252   
9991_ZBTB38_P1_ENSG00000177311            -0.022295         0.051294   
9992_ZBTB39_P1P2_ENSG00000166860           0.207843         0.118381   
9998_ZBTB44_P1P2_ENSG00000196323           0.019084        -0.108255   
9_AAR2_P1P2_ENSG00000131043                0.014528         0.280991   

gene_id                             ENSG00000160072  ENSG00000197785  ...  \
gene_transcript                                                       ...   
10004_ZBTB4_P2_ENSG00000174282            -0.091799         0.126250  ...   
10005_ZBTB4_P1_ENSG00

In [104]:
reduced_gene_name_index = [i.split("_")[1] for i in reduced_df.index.to_list()]
reduced_gene_promoter_index = [i.split('_')[2] for i in reduced_df.index.to_list()]
reduced_gene_name_and_promoter_index = [f'{i}_{j}' for i,j in zip(reduced_gene_name_index, reduced_gene_promoter_index)]
len(reduced_gene_name_and_promoter_index)

3758

In [108]:
np.unique(reduced_gene_name_and_promoter_index)

array(['AACS_P1P2', 'AAMDC_P1P2', 'AAMP_P1P2', ..., 'ZRSR2_P1P2',
       'ZW10_P1P2', 'ZWINT_P1P2'], shape=(3756,), dtype='<U77')

In [121]:
gene_name_with_promoter = SPG_df['gene_name_promoter'].to_list()

In [110]:
len(gene_name_with_promoter)

1973

In [122]:
indices_to_keep_based_on_name_and_promoter = []
j = 0
missing_gene_with_promoter_in_SPG = []
for ids, i in enumerate(reduced_gene_name_and_promoter_index): # this is from the dataframe.
    k = i.split('_')[0]
    if k in SPG_df['gene_name'].to_list():
        if i in gene_name_with_promoter: # this is from the SPG_df.
            indices_to_keep_based_on_name_and_promoter.append(ids)
        else:
            j += 1
            missing_gene_with_promoter_in_SPG.append(i)
    elif k in HVG_df['gene_name'].to_list():
        indices_to_keep_based_on_name_and_promoter.append(ids)
    else:
        print(i)

print(len(indices_to_keep_based_on_name_and_promoter))
print(j)

3630
128


In [117]:
SPG_df

gene_name gene_name_promoter        gene_code gene_promoter  cluster
0        ZBTB4           ZBTB4_P1  ENSG00000174282            P1       19
1        ZBTB5         ZBTB5_P1P2  ENSG00000168795          P1P2       -1
2      ZBTB8OS       ZBTB8OS_P1P2  ENSG00000176261          P1P2       -1
3       ZC3H13        ZC3H13_P1P2  ENSG00000123200          P1P2        4
4       ZC3H18        ZC3H18_P1P2  ENSG00000158545          P1P2       -1
...        ...                ...              ...           ...      ...
1968    ZBTB17        ZBTB17_P1P2  ENSG00000116809          P1P2        7
1969  C16orf87      C16orf87_P1P2  ENSG00000155330          P1P2        2
1970    ZBTB39        ZBTB39_P1P2  ENSG00000166860          P1P2       -1
1971    ZBTB44        ZBTB44_P1P2  ENSG00000196323          P1P2       -1
1972      AAR2          AAR2_P1P2  ENSG00000131043          P1P2       57

[1973 rows x 5 columns]

In [133]:
any(reduced_df.index.str.contains('ATXN7_ENST00000398590.3'))

True

In [127]:
#'ATXN7_ENST00000538065.1' 
'ATXN7' in SPG_df['gene_name'].to_list()

True

In [128]:
SPG_df[SPG_df['gene_name'] == 'ATXN7']

gene_name       gene_name_promoter        gene_code      gene_promoter  \
1327     ATXN7  ATXN7_ENST00000398590.3  ENSG00000163635  ENST00000398590.3   

      cluster  
1327       -1

In [116]:
len(missing_gene_with_promoter_in_SPG)

128

In [115]:
missing_gene_with_promoter_in_SPG

['ZBTB4_P2',
 'ZC3H4_P2',
 'ZMAT2_P2',
 'ZNF207_P2',
 'ZNF358_P2',
 'ZNF574_P2',
 'CBFA2T3_P1',
 'CCDC130_P2',
 'ADAT3_P1',
 'CDK6_P2',
 'CFDP1_P2',
 'CLK3_P2',
 'CNDP2_P2',
 'CPEB1_ENST00000562019.1',
 'AGBL5_P2',
 'DAXX_P2',
 'AHCYL1_P2',
 'DDB1_P2',
 'DDX46_P2',
 'DDX54_P2',
 'DHODH_P2',
 'DHX30_P2',
 'DNAAF3_P1',
 'DNASE1L1_P1',
 'DPF2_P2',
 'DPH1_P2',
 'EEF2_P2',
 'EGR1_P2',
 'EIF4G2_P2',
 'EIF5_P2',
 'EMC3_P2',
 'EPS8L1_P1',
 'ESF1_P2',
 'PRXL2A_P2',
 'FBXW7_ENST00000603841.1',
 'FOLR3_P1',
 'ABCF1_P2',
 'GRB2_P2',
 'HAUS7_P2',
 'HIC2_P2',
 'HIPK3_P2',
 'HIRA_P1',
 'AP2M1_P2',
 'HNRNPL_P2',
 'HNRNPM_P2',
 'HOXA3_P1',
 'IKZF2_P2',
 'KAT8_P2',
 'LDB1_P1',
 'LIME1_P1',
 'LIN54_P2',
 'LTBP4_P1',
 'LUC7L2_P2',
 'MAPKAP1_P2',
 'MAST1_P2',
 'MAT2B_P2',
 'MED15_P2',
 'MEIS3_P2',
 'MFN2_P2',
 'MIER1_P2',
 'MYCN_P2',
 'NCL_P2',
 'NDUFC1_P2',
 'NDUFS5_P2',
 'NFIC_P2',
 'PCBP2_P2',
 'PCID2_P2',
 'PELO_P2',
 'PHAX_P2',
 'PKM_P2',
 'PMPCB_P2',
 'PNISR_P2',
 'ATP5F1A_P2',
 'POLR1C_P2',
 'POLR1D

In [134]:
final_df = reduced_df.iloc[indices_to_keep_based_on_name_and_promoter].copy()
final_df


gene_id                             ENSG00000188976  ENSG00000187961  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282            -0.030201        -0.088615   
10006_ZBTB5_P1P2_ENSG00000168795          -0.094131         0.093585   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.042972        -0.122381   
10020_ZC3H13_P1P2_ENSG00000123200         -0.120841        -0.013023   
10022_ZC3H15_P1P2_ENSG00000065548         -0.069962         0.045348   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.023088         0.074530   
9991_ZBTB38_P1_ENSG00000177311             0.032109         0.040046   
9992_ZBTB39_P1P2_ENSG00000166860          -0.054780        -0.051475   
9998_ZBTB44_P1P2_ENSG00000196323           0.166343         0.202679   
9_AAR2_P1P2_ENSG00000131043                0.313340         0.215871   

gene_id                             ENSG00000187608  ENSG00000160087  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282             0.097081        -0.050281   
10006_ZBTB5_P1P2_ENSG00000168795           0.041780         0.011111   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.015089        -0.010220   
10020_ZC3H13_P1P2_ENSG00000123200          0.690666         0.369722   
10022_ZC3H15_P1P2_ENSG00000065548         -0.056278         0.054014   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.028234        -0.102966   
9991_ZBTB38_P1_ENSG00000177311             0.105273         0.070274   
9992_ZBTB39_P1P2_ENSG00000166860           0.206940         0.023562   
9998_ZBTB44_P1P2_ENSG00000196323          -0.026491        -0.038248   
9_AAR2_P1P2_ENSG00000131043                0.256190         0.451411   

gene_id                             ENSG00000127054  ENSG00000162576  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282             0.053923         0.047092   
10006_ZBTB5_P1P2_ENSG00000168795           0.049495         0.068986   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.021991        -0.140862   
10020_ZC3H13_P1P2_ENSG00000123200          0.229487         0.635588   
10022_ZC3H15_P1P2_ENSG00000065548          0.087813        -0.018570   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311             0.021527        -0.110965   
9991_ZBTB38_P1_ENSG00000177311             0.090874         0.093788   
9992_ZBTB39_P1P2_ENSG00000166860          -0.088279        -0.061373   
9998_ZBTB44_P1P2_ENSG00000196323           0.146278        -0.032702   
9_AAR2_P1P2_ENSG00000131043               -0.083404         0.057765   

gene_id                             ENSG00000175756  ENSG00000242485  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282             0.025221        -0.003499   
10006_ZBTB5_P1P2_ENSG00000168795          -0.062200         0.044283   
10015_ZBTB8OS_P1P2_ENSG00000176261        -0.131819         0.023055   
10020_ZC3H13_P1P2_ENSG00000123200          0.187934         0.145091   
10022_ZC3H15_P1P2_ENSG00000065548         -0.081548         0.034590   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.041697        -0.002252   
9991_ZBTB38_P1_ENSG00000177311            -0.022295         0.051294   
9992_ZBTB39_P1P2_ENSG00000166860           0.207843         0.118381   
9998_ZBTB44_P1P2_ENSG00000196323           0.019084        -0.108255   
9_AAR2_P1P2_ENSG00000131043                0.014528         0.280991   

gene_id                             ENSG00000160072  ENSG00000197785  ...  \
gene_transcript                                                       ...   
10005_ZBTB4_P1_ENSG00000174282            -0.085795        -0.027718  ...   
10006_ZBTB5_P1P2_ENSG

In [113]:
final_df.shape

(3630, 3460)

In [114]:
# are missing_gene_with_promoter_in_SPG in HVG
for i in missing_gene_with_promoter_in_SPG:
    k = i.split("_")[0]
    if k in HVG_df['gene_name']:
        continue
    else:
        print(f'{i} in HVG but not in SPG')

ZBTB4_P2 in HVG but not in SPG
ZC3H4_P2 in HVG but not in SPG
ZMAT2_P2 in HVG but not in SPG
ZNF207_P2 in HVG but not in SPG
ZNF358_P2 in HVG but not in SPG
ZNF574_P2 in HVG but not in SPG
CBFA2T3_P1 in HVG but not in SPG
CCDC130_P2 in HVG but not in SPG
ADAT3_P1 in HVG but not in SPG
CDK6_P2 in HVG but not in SPG
CFDP1_P2 in HVG but not in SPG
CLK3_P2 in HVG but not in SPG
CNDP2_P2 in HVG but not in SPG
CPEB1_ENST00000562019.1 in HVG but not in SPG
AGBL5_P2 in HVG but not in SPG
DAXX_P2 in HVG but not in SPG
AHCYL1_P2 in HVG but not in SPG
DDB1_P2 in HVG but not in SPG
DDX46_P2 in HVG but not in SPG
DDX54_P2 in HVG but not in SPG
DHODH_P2 in HVG but not in SPG
DHX30_P2 in HVG but not in SPG
DNAAF3_P1 in HVG but not in SPG
DNASE1L1_P1 in HVG but not in SPG
DPF2_P2 in HVG but not in SPG
DPH1_P2 in HVG but not in SPG
EEF2_P2 in HVG but not in SPG
EGR1_P2 in HVG but not in SPG
EIF4G2_P2 in HVG but not in SPG
EIF5_P2 in HVG but not in SPG
EMC3_P2 in HVG but not in SPG
EPS8L1_P1 in HVG but 

In [178]:
final_df.shape

(3630, 3460)

In [191]:
final_df

gene_id                             ENSG00000188976  ENSG00000187961  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282            -0.030201        -0.088615   
10006_ZBTB5_P1P2_ENSG00000168795          -0.094131         0.093585   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.042972        -0.122381   
10020_ZC3H13_P1P2_ENSG00000123200         -0.120841        -0.013023   
10022_ZC3H15_P1P2_ENSG00000065548         -0.069962         0.045348   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.023088         0.074530   
9991_ZBTB38_P1_ENSG00000177311             0.032109         0.040046   
9992_ZBTB39_P1P2_ENSG00000166860          -0.054780        -0.051475   
9998_ZBTB44_P1P2_ENSG00000196323           0.166343         0.202679   
9_AAR2_P1P2_ENSG00000131043                0.313340         0.215871   

gene_id                             ENSG00000187608  ENSG00000160087  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282             0.097081        -0.050281   
10006_ZBTB5_P1P2_ENSG00000168795           0.041780         0.011111   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.015089        -0.010220   
10020_ZC3H13_P1P2_ENSG00000123200          0.690666         0.369722   
10022_ZC3H15_P1P2_ENSG00000065548         -0.056278         0.054014   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.028234        -0.102966   
9991_ZBTB38_P1_ENSG00000177311             0.105273         0.070274   
9992_ZBTB39_P1P2_ENSG00000166860           0.206940         0.023562   
9998_ZBTB44_P1P2_ENSG00000196323          -0.026491        -0.038248   
9_AAR2_P1P2_ENSG00000131043                0.256190         0.451411   

gene_id                             ENSG00000127054  ENSG00000162576  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282             0.053923         0.047092   
10006_ZBTB5_P1P2_ENSG00000168795           0.049495         0.068986   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.021991        -0.140862   
10020_ZC3H13_P1P2_ENSG00000123200          0.229487         0.635588   
10022_ZC3H15_P1P2_ENSG00000065548          0.087813        -0.018570   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311             0.021527        -0.110965   
9991_ZBTB38_P1_ENSG00000177311             0.090874         0.093788   
9992_ZBTB39_P1P2_ENSG00000166860          -0.088279        -0.061373   
9998_ZBTB44_P1P2_ENSG00000196323           0.146278        -0.032702   
9_AAR2_P1P2_ENSG00000131043               -0.083404         0.057765   

gene_id                             ENSG00000175756  ENSG00000242485  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282             0.025221        -0.003499   
10006_ZBTB5_P1P2_ENSG00000168795          -0.062200         0.044283   
10015_ZBTB8OS_P1P2_ENSG00000176261        -0.131819         0.023055   
10020_ZC3H13_P1P2_ENSG00000123200          0.187934         0.145091   
10022_ZC3H15_P1P2_ENSG00000065548         -0.081548         0.034590   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.041697        -0.002252   
9991_ZBTB38_P1_ENSG00000177311            -0.022295         0.051294   
9992_ZBTB39_P1P2_ENSG00000166860           0.207843         0.118381   
9998_ZBTB44_P1P2_ENSG00000196323           0.019084        -0.108255   
9_AAR2_P1P2_ENSG00000131043                0.014528         0.280991   

gene_id                             ENSG00000160072  ENSG00000197785  ...  \
gene_transcript                                                       ...   
10005_ZBTB4_P1_ENSG00000174282            -0.085795        -0.027718  ...   
10006_ZBTB5_P1P2_ENSG

In [192]:
adata

AnnData object with n_obs × n_vars = 11258 × 8248
    obs: 'UMI_count_unfiltered', 'num_cells_unfiltered', 'num_cells_filtered', 'control_expr', 'fold_expr', 'pct_expr', 'core_control', 'mean_leverage_score', 'std_leverage_score', 'energy_test_p_value', 'anderson_darling_counts', 'mann_whitney_counts', 'z_gemgroup_UMI', 'mitopercent', 'TE_ratio', 'cnv_score_z'
    var: 'gene_name', 'mean', 'std', 'cv', 'in_matrix', 'gini', 'clean_mean', 'clean_std', 'clean_cv'

In [206]:
final_df

gene_id                             ENSG00000188976  ENSG00000187961  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282            -0.030201        -0.088615   
10006_ZBTB5_P1P2_ENSG00000168795          -0.094131         0.093585   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.042972        -0.122381   
10020_ZC3H13_P1P2_ENSG00000123200         -0.120841        -0.013023   
10022_ZC3H15_P1P2_ENSG00000065548         -0.069962         0.045348   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.023088         0.074530   
9991_ZBTB38_P1_ENSG00000177311             0.032109         0.040046   
9992_ZBTB39_P1P2_ENSG00000166860          -0.054780        -0.051475   
9998_ZBTB44_P1P2_ENSG00000196323           0.166343         0.202679   
9_AAR2_P1P2_ENSG00000131043                0.313340         0.215871   

gene_id                             ENSG00000187608  ENSG00000160087  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282             0.097081        -0.050281   
10006_ZBTB5_P1P2_ENSG00000168795           0.041780         0.011111   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.015089        -0.010220   
10020_ZC3H13_P1P2_ENSG00000123200          0.690666         0.369722   
10022_ZC3H15_P1P2_ENSG00000065548         -0.056278         0.054014   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.028234        -0.102966   
9991_ZBTB38_P1_ENSG00000177311             0.105273         0.070274   
9992_ZBTB39_P1P2_ENSG00000166860           0.206940         0.023562   
9998_ZBTB44_P1P2_ENSG00000196323          -0.026491        -0.038248   
9_AAR2_P1P2_ENSG00000131043                0.256190         0.451411   

gene_id                             ENSG00000127054  ENSG00000162576  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282             0.053923         0.047092   
10006_ZBTB5_P1P2_ENSG00000168795           0.049495         0.068986   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.021991        -0.140862   
10020_ZC3H13_P1P2_ENSG00000123200          0.229487         0.635588   
10022_ZC3H15_P1P2_ENSG00000065548          0.087813        -0.018570   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311             0.021527        -0.110965   
9991_ZBTB38_P1_ENSG00000177311             0.090874         0.093788   
9992_ZBTB39_P1P2_ENSG00000166860          -0.088279        -0.061373   
9998_ZBTB44_P1P2_ENSG00000196323           0.146278        -0.032702   
9_AAR2_P1P2_ENSG00000131043               -0.083404         0.057765   

gene_id                             ENSG00000175756  ENSG00000242485  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282             0.025221        -0.003499   
10006_ZBTB5_P1P2_ENSG00000168795          -0.062200         0.044283   
10015_ZBTB8OS_P1P2_ENSG00000176261        -0.131819         0.023055   
10020_ZC3H13_P1P2_ENSG00000123200          0.187934         0.145091   
10022_ZC3H15_P1P2_ENSG00000065548         -0.081548         0.034590   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.041697        -0.002252   
9991_ZBTB38_P1_ENSG00000177311            -0.022295         0.051294   
9992_ZBTB39_P1P2_ENSG00000166860           0.207843         0.118381   
9998_ZBTB44_P1P2_ENSG00000196323           0.019084        -0.108255   
9_AAR2_P1P2_ENSG00000131043                0.014528         0.280991   

gene_id                             ENSG00000160072  ENSG00000197785  ...  \
gene_transcript                                                       ...   
10005_ZBTB4_P1_ENSG00000174282            -0.085795        -0.027718  ...   
10006_ZBTB5_P1P2_ENSG

In [217]:
final_df.to_csv('union_genes_clustered_df.csv', index_label='gene_transcript')

In [220]:
t = pd.read_csv('data/union_genes_clustered_df.csv', index_col=0)
t

ENSG00000188976  ENSG00000187961  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282            -0.030201        -0.088615   
10006_ZBTB5_P1P2_ENSG00000168795          -0.094131         0.093585   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.042972        -0.122381   
10020_ZC3H13_P1P2_ENSG00000123200         -0.120841        -0.013023   
10022_ZC3H15_P1P2_ENSG00000065548         -0.069962         0.045348   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.023088         0.074530   
9991_ZBTB38_P1_ENSG00000177311             0.032109         0.040046   
9992_ZBTB39_P1P2_ENSG00000166860          -0.054780        -0.051475   
9998_ZBTB44_P1P2_ENSG00000196323           0.166343         0.202679   
9_AAR2_P1P2_ENSG00000131043                0.313340         0.215871   

                                    ENSG00000187608  ENSG00000160087  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282             0.097081        -0.050281   
10006_ZBTB5_P1P2_ENSG00000168795           0.041780         0.011111   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.015089        -0.010220   
10020_ZC3H13_P1P2_ENSG00000123200          0.690666         0.369722   
10022_ZC3H15_P1P2_ENSG00000065548         -0.056278         0.054014   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.028234        -0.102966   
9991_ZBTB38_P1_ENSG00000177311             0.105273         0.070274   
9992_ZBTB39_P1P2_ENSG00000166860           0.206940         0.023562   
9998_ZBTB44_P1P2_ENSG00000196323          -0.026491        -0.038248   
9_AAR2_P1P2_ENSG00000131043                0.256190         0.451411   

                                    ENSG00000127054  ENSG00000162576  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282             0.053923         0.047092   
10006_ZBTB5_P1P2_ENSG00000168795           0.049495         0.068986   
10015_ZBTB8OS_P1P2_ENSG00000176261         0.021991        -0.140862   
10020_ZC3H13_P1P2_ENSG00000123200          0.229487         0.635588   
10022_ZC3H15_P1P2_ENSG00000065548          0.087813        -0.018570   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311             0.021527        -0.110965   
9991_ZBTB38_P1_ENSG00000177311             0.090874         0.093788   
9992_ZBTB39_P1P2_ENSG00000166860          -0.088279        -0.061373   
9998_ZBTB44_P1P2_ENSG00000196323           0.146278        -0.032702   
9_AAR2_P1P2_ENSG00000131043               -0.083404         0.057765   

                                    ENSG00000175756  ENSG00000242485  \
gene_transcript                                                        
10005_ZBTB4_P1_ENSG00000174282             0.025221        -0.003499   
10006_ZBTB5_P1P2_ENSG00000168795          -0.062200         0.044283   
10015_ZBTB8OS_P1P2_ENSG00000176261        -0.131819         0.023055   
10020_ZC3H13_P1P2_ENSG00000123200          0.187934         0.145091   
10022_ZC3H15_P1P2_ENSG00000065548         -0.081548         0.034590   
...                                             ...              ...   
9990_ZBTB38_P2_ENSG00000177311            -0.041697        -0.002252   
9991_ZBTB38_P1_ENSG00000177311            -0.022295         0.051294   
9992_ZBTB39_P1P2_ENSG00000166860           0.207843         0.118381   
9998_ZBTB44_P1P2_ENSG00000196323           0.019084        -0.108255   
9_AAR2_P1P2_ENSG00000131043                0.014528         0.280991   

                                    ENSG00000160072  ENSG00000197785  ...  \
gene_transcript                                                       ...   
10005_ZBTB4_P1_ENSG00000174282            -0.085795        -0.027718  ...   
10006_ZBTB5_P1P2_ENSG00000168795          -0.089965      